doc/design/async_devices.md

```diff
@@ -0,0 +1,68 @@
+# Design Doc: Asynchronous and Non-blocking techniques on heterogeneous devices.
+
+
+## Problem
+
+We often use heterogeneous devices, like GPU, FPGA, in a deep learning platform. The computation on heterogeneous devices is usually asynchronous and non-blocking. The computation tasks are offloaded to heterogeneous devices and programmers must wait for the computation finished and then fetch results. Heterogeneous devices are usually able to execute simultaneously. For example, Fermi architecture (CUDA compute capability 2.0+) can simultaneously support
+
+* execute CUDA kernels
+* host-to-device memory copy
+* device-to-host memory copy.
+
+However, the simultaneous execution is not transparent to programmers. 
+
+Let's use CUDA as an example. There is a building block named `stream` in CUDA. Streams introduce task-based parallelism to CUDA codes. The sequence of operations will be executed in issue-order on the GPU if they are in the same stream. 
+
+The operators in different streams are able to run concurrently as long as they are in multiple streams and hardware supports it. CUDA hardware has no notion of streams. The hardware has separate queues (engines) to perform memory copies and to execute kernels.
+
+If we want to take advantage of CUDA devices, we must use at least N streams, where N equals the number of hardware queues, and separate operators into these streams. The N equals to three since CUDA can simultaneously execute CUDA kernels, H2D memcpy, D2H memcpy by the CUDA hardware.
```

Comment:
It seems that the stream of CUDA does not have a limit. As long as the resources (memory and computing) of GPU are not occupied, in theory, you can create a new stream.

Reply:
No, it does not. The CUDA can create many streams without any limits. However, jobs in CUDA can simultaneously execute, in two conditions.

1. The jobs in different streams.
2. The hardware supports.

Since the CUDA hardware supports to simultaneously execute THREE kinds of operators, `Kernel execution`/`D2H memcpy`/`H2D memcpy`, we need AT LEAST THREE streams to make full usage of CUDA devices.

Reply:
Threads in a block will be launched in a SM(streaming multiprocessors).  If the former CUDA kernel occupies few SMs and there is more SM left, another CUDA kernel can be executed in parallel. Please refer to https://devblogs.nvidia.com/gpu-pro-tip-cuda-7-streams-simplify-concurrency/.

Reply:
"And if we use more than three streams, it will not help much since the hardware only supports to simultaneously execute THREE kinds of operators."

I don't agree with this statement. Consider the following case:

We have 7 kernels with **no dependency** with each other. If we only have 3 streams, it's possible that they are put into 3 stream in the following order:

A, B, C, D, E
F
G

If A takes 100s to run, and the rest takes 1s to run. Then B, C, D, E would unnecessarily wait 100s for A to complete. If we place them into 7 independent streams, B, C, D, E, F, G would complete within 3s in the worst case when there are 3 hardware queues.

I think the number of streams we use should equal to the concurrency expressed in our program, not the hardware.